c/datasetjson.c

```diff
@@ -1658,52 +1658,35 @@ void respondWithDatasetMetadata(HttpResponse *response) {
     respondWithError(response,HTTP_STATUS_BAD_REQUEST,"No dataset name given");
     return;
   }
-
   char *username = response->request->username;
   int dsnLen = strlen(datasetOrMember);
+  char *percentDecoded = cleanURLParamValue(response->slh, datasetOrMember);
+  char *absDsPath_temp = stringConcatenate(response->slh, "//'", percentDecoded);
+  char *absDsPath = stringConcatenate(response->slh, absDsPath_temp, "'");
+
+  if(!isDatasetPathValid(absDsPath)){
+    respondWithError(response,HTTP_STATUS_BAD_REQUEST,"Invalid dataset path");
+    return;
+  }
+
   int lParenIndex = indexOf(datasetOrMember, dsnLen, '(', 0);
-  char pdsDSN[45];
-  char pdsMember[100];
+  DatasetName dsnName;
+  DatasetMemberName memName;
+  extractDatasetAndMemberName(absDsPath, &dsnName, &memName);
   int memberNameLength = 0;
+
   if (lParenIndex > 0){
-    int rParenIndex = lastIndexOf(datasetOrMember, dsnLen, ')');
-    int encodedMemberNameLength = rParenIndex-lParenIndex;
```

Comment:
I think you could extend the dataset and member structure. You could a `uint8_t` field with the length of the value. That field would be set by the `extractDatasetAndMemberName` function so that you don't need to have for-loops here.